### ISSUE TITLE
Add XChaCha/XSalsa support

### ISSUE THREAD

**Issue by jeremyherbert** [Mbed-TLS/mbedtls#3381]
### Description
- Type: Enhancement\Feature Request
- Priority: Minor

----------------------------------------------------------------
## Enhancement\Feature Request

The main problem with the two most common AEAD ciphers (AES-GCM and ChaCha20-Poly1305) is that nonce reuse is catastrophic. These two ciphers are not ideal for random nonces, as they are limited to 96 bit nonces. In a scenario with long-term keys and using random nonces, this means that the probability of a birthday attack starts to get uncomfortable for a large number of messages. A more appropriate design would be to use a monotonically increasing counter as the nonce, but a significant proportion of embedded devices do not have a high-endurance non-volatile storage mechanism (let alone one which is secure).

XChaCha increases the nonce size to 192 bits, which allows for safe use of a random nonce. As far more devices have HWRNGs or a source of entropy than high endurance RW storage, and as mbedtls is primarily targeted at embedded devices, it would be great to have support for this in mbedtls. It is already available in libsodium and other popular cryptography libraries.

RFC: https://datatracker.ietf.org/doc/draft-irtf-cfrg-xchacha/

**Comment by jeremyherbert** [comment]
@gilles-peskine-arm just looking for a bit of guidance on how to implement this so it fits into the existing structure of mbedtls. 

- The implementation would need access to some of the static methods in `chacha20.c` (the `chacha20_block` function). With this in mind, should it be in implemented in new files `xchacha20.c` and `xchachapoly.c`, or included in the existing `chacha20.c` and `chachapoly.c`? I think probably the latter.
- If it is inside the existing files, should there be a preprocessor define to allow it to be skipped?

**Comment by gilles-peskine-arm** [comment]
Thanks for volunteering! Whether to put the code in the same file, and whether to have a preprocessor symbol, depends on how much code is shared vs specific. I'm not familiar enough with (X)ChaCha to guess at the numbers.

Given that some of the code is shared, and two modules are affected, I think it's best to put the code in the same files, and see how much larger they grow. If it's too much, add config symbols for the specific algorithms, like `rsa.c` has parts of it gated by `MBEDTLS_PKCS1_V15` and `MBEDTLS_PKCS1_V21`.

Please note that our [bandwidth](https://github.com/ARMmbed/mbedtls/projects/4) for reviews in June is already full, and given the backlog July isn't looking good either. So I'm afraid it might take a while until we can review a new cryptographic algorithm.

**Comment by jeremyherbert** [comment]
Thanks for the guidance. I'll try to put it together inside the existing files and it can be refactored later if necessary.

No problems about the review delay; it's actually mostly similar to chacha20 anyway, so once you get around to it, hopefully it should be fairly straightforward.

**Comment by inorick** [comment]
@jeremyherbert Did you ever get around to implement this?

**Comment by jeremyherbert** [comment]
Sorry, I did not. Ended up becoming out of scope for the project I was working on, and I haven't had a chance to come back to it yet.

**Comment by daverodgman** [comment]
Closing - duplicate of #6557
